util/app/discovery/discovery.go

```diff
@@ -77,67 +77,47 @@ func AppType(ctx context.Context, path string, enableGenerateManifests map[strin
 	return "Directory", nil
 }
 
-// 1. list all plugins in /plugins folder
-// 2. foreach plugin setup connection with respective cmp-server
-// 3. check isSupported(path)?
-// 4.a if no then close connection
-// 4.b if yes then return conn for detected plugin
-func DetectConfigManagementPlugin(ctx context.Context, repoPath string, env []string, tarExcludedGlobs []string) (io.Closer, pluginclient.ConfigManagementPluginServiceClient, error) {
+// if pluginName is provided setup connection to that cmp-server
+// else
+// list all plugins in /plugins folder and foreach plugin
+// check cmpSupports()
+// if supported return conn for the cmp-server
+
+func DetectConfigManagementPlugin(ctx context.Context, repoPath, pluginName string, env []string, tarExcludedGlobs []string) (io.Closer, pluginclient.ConfigManagementPluginServiceClient, error) {
 	var conn io.Closer
 	var cmpClient pluginclient.ConfigManagementPluginServiceClient
+	var connFound bool
 
 	pluginSockFilePath := common.GetPluginSockFilePath()
 	log.WithFields(log.Fields{
 		common.SecurityField:    common.SecurityLow,
 		common.SecurityCWEField: 775,
 	}).Debugf("pluginSockFilePath is: %s", pluginSockFilePath)
 
-	files, err := os.ReadDir(pluginSockFilePath)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	var connFound bool
-	for _, file := range files {
-		if file.Type() == os.ModeSocket {
-			address := filepath.Join(pluginSockFilePath, file.Name())
-			cmpclientset := pluginclient.NewConfigManagementPluginClientSet(address)
-
-			conn, cmpClient, err = cmpclientset.NewConfigManagementPluginClient()
-			if err != nil {
-				log.WithFields(log.Fields{
-					common.SecurityField:    common.SecurityMedium,
-					common.SecurityCWEField: 775,
-				}).Errorf("error dialing to cmp-server for plugin %s, %v", file.Name(), err)
-				continue
-			}
-
-			isSupported, err := matchRepositoryCMP(ctx, repoPath, cmpClient, env, tarExcludedGlobs)
-			if err != nil {
-				log.WithFields(log.Fields{
-					common.SecurityField:    common.SecurityMedium,
-					common.SecurityCWEField: 775,
-				}).Errorf("repository %s is not the match because %v", repoPath, err)
-				continue
-			}
-
-			if !isSupported {
-				log.WithFields(log.Fields{
-					common.SecurityField:    common.SecurityLow,
-					common.SecurityCWEField: 775,
-				}).Debugf("Reponse from socket file %s is not supported", file.Name())
-				io.Close(conn)
-			} else {
-				connFound = true
-				break
+	if pluginName != "" {
+		// check if the given plugin supports the repo
+		conn, cmpClient, connFound = cmpSupports(ctx, pluginSockFilePath, repoPath, fmt.Sprintf("%v.sock", pluginName), env, tarExcludedGlobs)
+		if !connFound {
+			return nil, nil, fmt.Errorf("Couldn't find cmp-server plugin with name %v supporting repository %s", pluginName, repoPath)
+		}
+	} else {
+		files, err := os.ReadDir(pluginSockFilePath)
+		if err != nil {
+			return nil, nil, err
```

Comment:
Can we wrap this error message with fmt.Errorf?

Reply:
Sure.